src/interface/swarm_pack_base.hpp

```diff
@@ -0,0 +1,267 @@
+  }
+
+  // Actually build a `SwarmPackBase` (i.e. create a view of views, fill on host, and
+  // deep copy the view of views to device) from the variables specified in desc contained
+  // from the blocks contained in pmd (which can either be MeshBlockData/MeshData).
+  template <class T>
+  static SwarmPackBase<TYPE> Build(T *pmd, const SwarmPackDescriptor<TYPE> &desc) {
+    int nvar = desc.vars.size();
+
+    SwarmPackBase<TYPE> pack;
+    pack.nvar_ = desc.vars.size();
+
+    // Count up the size of the array that is required
+    int max_size = 0;
+    int nblocks = 0;
+    int ndim = 3;
+    std::vector<int> vardims;
+    ForEachBlock(pmd, std::vector<bool>{}, [&](int b, auto *pmbd) {
+      auto swarm = pmbd->GetSwarm(desc.swarm_name);
+      int size = 0;
+      nblocks++;
+      for (auto &pv : swarm->template GetVariableVector<TYPE>()) {
+        for (int i = 0; i < nvar; ++i) {
+          if (desc.IncludeVariable(i, pv)) {
+            size += pv->GetDim(6) * pv->GetDim(5) * pv->GetDim(4) * pv->GetDim(3) *
+                    pv->GetDim(2);
+            ndim = 1;
+            int vardim = (pv->GetDim(6) > 1 ? 1 : 0) + (pv->GetDim(5) > 1 ? 1 : 0) +
+                         (pv->GetDim(4) > 1 ? 1 : 0) + (pv->GetDim(3) > 1 ? 1 : 0) +
+                         (pv->GetDim(2) > 1 ? 1 : 0);
+            vardims.push_back(vardim);
+          }
+        }
+      }
+      max_size = std::max(size, max_size);
+    });
+    pack.nblocks_ = nblocks;
+
+    // Allocate the views
+    int leading_dim = 1;
+    pack.pack_ = pack_t("data_ptr", leading_dim, nblocks, max_size);
+    auto pack_h = Kokkos::create_mirror_view(pack.pack_);
+
+    int bounds_leading_size = 2 + 1 + *max_element(vardims.begin(), vardims.end());
+    pack.bounds_ = bounds_t("bounds", bounds_leading_size, nblocks, nvar);
+    auto bounds_h = Kokkos::create_mirror_view(pack.bounds_);
+
+    // Fill the views
+    ForEachBlock(pmd, std::vector<bool>{}, [&](int b, auto *pmbd) {
+      int idx = 0;
+      auto swarm = pmbd->GetSwarm(desc.swarm_name);
+      for (int i = 0; i < nvar; ++i) {
+        bounds_h(0, b, i) = idx;
+        for (auto &pv : swarm->template GetVariableVector<TYPE>()) {
+          if (desc.IncludeVariable(i, pv)) {
+            for (int t = 0; t < pv->GetDim(6); ++t) {
+              for (int u = 0; u < pv->GetDim(5); ++u) {
+                for (int v = 0; v < pv->GetDim(4); ++v) {
+                  for (int l = 0; l < pv->GetDim(3); ++l) {
+                    for (int m = 0; m < pv->GetDim(2); ++m) {
+                      pack_h(0, b, idx) = pv->data.Get(0, t, u, v, l, m);
+                      PARTHENON_REQUIRE(
+                          pack_h(0, b, idx).size() > 0,
+                          "Seems like this variable might not actually be allocated.");
+                      idx++;
+                    }
+                  }
+                }
+              }
+            }
+            bounds_h(2, b, i) = (pv->GetDim(6) > 1 ? 1 : 0) +
```

Comment:
Maybe a brief comment that you are storing the tensor size of the swarm variable in `bounds`? At least I think that is what is being done here.

Reply:
It's actually not clear to me what the bounds calls/array does for swarms. Do we need one?

Reply:
I removed this extraneous debugging info from bounds.  Thanks for pointing me to additional cleanup.